server/src/main/java/org/opensearch/cluster/routing/IndexShardRoutingTable.java

```diff
@@ -292,6 +292,86 @@ public ShardIterator activeInitializingShardsRankedIt(
         return new PlainShardIterator(shardId, ordered);
     }
 
+    /**
+     * Returns an iterator over active and initializing shards, shards are ordered by weighted
+     * round-robin scheduling policy.
+     *
+     * @param weightedRouting entity
+     * @param nodes           discovered nodes in the cluster
+     * @return an iterator over active and initializing shards, ordered by weighted round-robin
+     * scheduling policy. Making sure that initializing shards are the last to iterate through.
+     */
+    public ShardIterator activeInitializingShardsWeightedIt(
+        WeightedRouting weightedRouting,
+        DiscoveryNodes nodes,
+        WeightedRoutingCache cache,
+        int defaultWeight
+    ) {
+        final int seed = shuffler.nextSeed();
+        List<ShardRouting> ordered = new ArrayList<>();
+        List<ShardRouting> orderedActiveShards = cache.get(new WeightedRoutingCache.Key(shardId));
+        if (orderedActiveShards == null) {
+            orderedActiveShards = shardsOrderedByWeight(activeShards, weightedRouting, nodes, defaultWeight);
+            cache.put(new WeightedRoutingCache.Key(shardId), orderedActiveShards);
+        }
+        ordered.addAll(shuffler.shuffle(orderedActiveShards, seed));
+
+        if (!allInitializingShards.isEmpty()) {
+            List<ShardRouting> orderedInitializingShards = shardsOrderedByWeight(
+                allInitializingShards,
+                weightedRouting,
+                nodes,
+                defaultWeight
+            );
+            ordered.addAll(orderedInitializingShards);
+        }
+        return new PlainShardIterator(shardId, ordered);
+    }
```

Comment:
Took a second look, we already have a cached copy of shard and its attribute as referenced below
https://github.com/opensearch-project/OpenSearch/blob/3aef125d0dbf2f29d9bf00b84cb52892a27a5328/server/src/main/java/org/opensearch/cluster/routing/IndexShardRoutingTable.java#L577-L588

Do you think we can leverage the same and modify it to update on weight changes

Reply:
The advantage is this cache is re-built proactively and not on the request path, which would avoid any performance regression

Reply:
Created an [issue](https://github.com/opensearch-project/OpenSearch/issues/4525) for this

Reply:
Removed WeightedRoutingCache and created a map in IndexShardRoutingTable to store routings in memory until weights or shard routing table changes.